### PR TITLE
Add noIntroOfferExists as an IntroEligibilityStatus

### PR DESCRIPTION
**Pull request by joshdholtz**
### Checklist
- [x] If applicable, unit tests
- [ ] If applicable, create follow-up issues for `purchases-android` and hybrids

### Motivation
Fixes [sc-10961]
Intro offer eligibility was only using `eligible` or `ineligible`. This meant that `ineligible` was even being used for products that didn't have an intro offer attached to it. This PR adds `noIntroOfferExists` for those cases.

👟 This also stomps a bug where `productsWithIntroOffers` was filtering out products that didn't have an intro offer. This was bug was introduced in v4.

### Description
- Added new `IntroEligibilityStatus. noIntroOfferExists`
- StoreKit 1 (on device)
  - Renamed `productsWithIntroOffers` to `productsWithKnownIntroEligibilityStatus` 
    - Maps products no intro offer to `noIntroOfferExists`
  - Reworked `getIntroEligibility` in [this commit](https://github.com/RevenueCat/purchases-ios/pull/1267/commits/8f24508aad9fedb827075459a44c4f15631860b6) to...
    - Step 1: Filter out products without introductory discount and give .noIntroOfferExists status
    - Step 2: Send products without eligibility status to backend
    - Step 3: Merge results from step 1 and step 2
- StoreKit 2 (on device)
  - Updates `TrialOrIntroPriceEligibilityChecker/sk2CheckEligibility` to return `. noIntroOfferExists` if `sk2Product.subscription` is `nil`

### CODE REVIEW

File: Purchases/Purchasing/IntroEligibilityCalculator.swift [hunk 1]
```diff
@@ -94,10 +95,13 @@ private extension IntroEligibilityCalculator {
                     return foundByGroupId
                 }
 
-            let hasIntroductoryPrice = candidate.introductoryDiscount != nil
-            result[candidate.productIdentifier] = !hasIntroductoryPrice || usedIntroForProductIdentifier
-                ? IntroEligibilityStatus.ineligible
-                : IntroEligibilityStatus.eligible
+            if candidate.introductoryDiscount == nil {
+                result[candidate.productIdentifier] = .noIntroOfferExists
```
Comment (joshdholtz) [inline Purchases/Purchasing/IntroEligibilityCalculator.swift:99]: >If you've set up introductory prices in App Store Connect, the introductory price property will be populated. This property is nil if the product has no introductory price.

Taken from https://developer.apple.com/documentation/storekit/skproduct/2936878-introductoryprice

File: Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift [hunk 2]
```diff
@@ -88,14 +88,14 @@ class TrialOrIntroPriceEligibilityChecker {
...
 
             let eligibilityStatus: IntroEligibilityStatus
 
-            if let isEligible = isEligible {
+            if let subscription = sk2Product.subscription {
+                let isEligible = await subscription.isEligibleForIntroOffer
                 eligibilityStatus = isEligible ? .eligible : .ineligible
             } else {
-                eligibilityStatus = .unknown
+                eligibilityStatus = .noIntroOfferExists
```
Comment (joshdholtz) [inline Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift:98]:
```swift
guard let renewableSubscription = product.subscription else {
        // No renewable subscription is available for this product.
        return false
    }
```

Taken from https://developer.apple.com/documentation/storekit/product/subscriptioninfo/3803203-iseligibleforintrooffer#

File: Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift [hunk 3]
```diff
@@ -153,14 +153,22 @@ fileprivate extension TrialOrIntroPriceEligibilityChecker {
...
             let eligibility: [(String, IntroEligibility)] = Array(products.value ?? [])
-                .filter { $0.introductoryDiscount != nil }
-                .map { ($0.productIdentifier, IntroEligibility(eligibilityStatus: .eligible)) }
+                .map({ storeProduct in
+                    let status: IntroEligibilityStatus
+                    if storeProduct.introductoryDiscount == nil {
+                        status = .noIntroOfferExists
+                    } else {
+                        status = .eligible
+                    }
```
Comment (joshdholtz) [inline Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift:169]: This seems to be used for StoreKit 1 with iOS 11.2 but below iOS 12... unless there is an error in iOS 12 and up and then this is used as a secondary check

File: Purchases/Purchasing/IntroEligibilityCalculator.swift (hunk 1 above)
Comment (taquitos) [reply Purchases/Purchasing/IntroEligibilityCalculator.swift:99]: Thank you for pointing this documentation out 😄

File: Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift [hunk 4]
```diff
@@ -153,14 +153,22 @@ fileprivate extension TrialOrIntroPriceEligibilityChecker {
...
 
     @available(iOS 11.2, macOS 10.13.2, macCatalyst 13.0, tvOS 11.2, watchOS 6.2, *)
-    func productsWithIntroOffers(productIdentifiers: [String], completion: @escaping ReceiveIntroEligibilityBlock) {
+    func mapProductsWithIntroEligibility(productIdentifiers: [String],
+                                         completion: @escaping ReceiveIntroEligibilityBlock) {
         self.productsManager.products(withIdentifiers: Set(productIdentifiers)) { products in
             let eligibility: [(String, IntroEligibility)] = Array(products.value ?? [])
-                .filter { $0.introductoryDiscount != nil }
-                .map { ($0.productIdentifier, IntroEligibility(eligibilityStatus: .eligible)) }
+                .map({ storeProduct in
```
Comment (taquitos) [inline Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift:163]: `.map({` -> `map {`

**Review by taquitos — APPROVED** [review]
Style nit but looks great!
🎈🐐

File: Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift (hunk 4 above)
Comment (joshdholtz) [reply Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift:163]: I'm mad the linter didn't catch this 🙄

Comment (NachoSoto) [reply Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift:163]: There's probably a rule for it

File: StoreKitUnitTests/TrialOrIntroPriceEligibilityCheckerSK1Tests.swift [hunk 5]
```diff
@@ -75,6 +75,35 @@ class TrialOrIntroPriceEligibilityCheckerSK1Tests: StoreKitConfigTestCase {
         expect(receivedEligibilities.count) == 1
     }
 
+    func testSK1EligibilityMapAllProductsWithIntroEligibility() throws {
```
Comment (aboedo) [inline StoreKitUnitTests/TrialOrIntroPriceEligibilityCheckerSK1Tests.swift:78]: this test would fail right now in `main`, right?

**Review by aboedo — APPROVED** [review]

**Comment by aboedo** [comment]
thinking about this more, maybe we should actually turn this into a bigger thing and fix backend and sk2 implementations before shipping. 

Otherwise, we'll have an inconsistent API that, given the same products, will return an entirely different output depending on the OS version that's running and whether the local receipt could be parsed, which seems problematic.

**Comment by aboedo** [comment]
I think we could still take care of the backend one, though: we're already checking if the products have intro pricing before making the backend request, so we can take that opportunity to map them. 
See https://github.com/RevenueCat/purchases-ios/blob/main/Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift#L174

File: Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift (hunk 4 above)
Comment (NachoSoto) [reply Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift]: This is the rule: https://realm.github.io/SwiftLint/trailing_closure.html
Unfortunately it forces using the trailing closure syntax everywhere, which we probably don't want (it's nice to be explicit with the parameter name sometimes).

File: StoreKitUnitTests/TrialOrIntroPriceEligibilityCheckerSK1Tests.swift (hunk 5 above)
Comment (joshdholtz) [reply StoreKitUnitTests/TrialOrIntroPriceEligibilityCheckerSK1Tests.swift:78]: Yup yup! I wrote this test and commented out my fixed and it failed. Reapplied my fix and then it passed 😊

**Comment by joshdholtz** [comment]
> I think we could still take care of the backend one, though: we're already checking if the products have intro pricing before making the backend request, so we can take that opportunity to map them.
> See https://github.com/RevenueCat/purchases-ios/blob/main/Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift#L174

@aboedo I think I'm confused by this comment but also by what the backend does 😅

It looks like the backend sent product identifiers that:
1) that have no introductory price status (because we've already figured it out)
2) product identifiers that both may or may not have an introductory price (we can't be certain due to older OSes)

So any errors in the backend response might be because:
1) Not eligible for introductory pricing
2) No introductory price is attached to that produce

So any error coming from the backend needs to be an `.unknown`, correct?

**Comment by aboedo** [comment]
right, what I mean is that currently (correct me if I'm wrong), when checking intro eligibility in a situation where we need to use the backend implementation, we do: 

```
get products that have intro prices
send them to backend
map the backend response to either unknown, eligible or ineligible
map the ones we didn't send to the backend (because they didn't have intro prices) to unknown
```

That last bit is where we could map them to `.noIntroOffer`

**Comment by joshdholtz** [comment]
@aboedo Ahh, that is what this does 👉 https://github.com/RevenueCat/purchases-ios/blob/add-noIntroOfferExists-for-intro-eligibility-check/Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift#L136-L154. There are two different approaches to backend...

## If iOS 11.2 is available 
1. Get products and set status to `. noIntroOfferExists` in no intro offer  in `mapProductsWithIntroEligibility ` (https://github.com/RevenueCat/purchases-ios/blob/add-noIntroOfferExists-for-intro-eligibility-check/Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift#L159-L176)

😱  Which I realized has a bug in it because it also marks things as `.eligible` which we don't actually know so I need to fix that

2. Then sends products to backend and only fetches products that don't have a status (meaning products found with `.noIntroOfferExists` in step 1 are skipped)

## If under iOS 11.2
1. We just send up everything to backend because we can't do anything locally 🤷‍♂️ 

---

TL;DR - This change maps to `. noIntroOfferExists` before going to backend because backend filtering logic would already prevent these from being check because they have status

**Review by taquitos — APPROVED** [review]
This seems reasonable.
I'd need to sit down and recall how this works in order to give a solid thumbs up, but nothing here surprises me.

File: Purchases/Purchasing/IntroEligibilityCalculator.swift [hunk 6]
```diff
@@ -30,6 +30,7 @@ class IntroEligibilityCalculator {
     func checkEligibility(with receiptData: Data,
                           productIdentifiers candidateProductIdentifiers: Set<String>,
                           completion: @escaping ([String: IntroEligibilityStatus], Error?) -> Void) {
+
```
Comment (NachoSoto) [inline Purchases/Purchasing/IntroEligibilityCalculator.swift:33]: Nit: we only add spaces to types not functions

Comment (joshdholtz) [reply Purchases/Purchasing/IntroEligibilityCalculator.swift:33]: GitHub needs a 🤦‍♂️ reaction

**Review by aboedo — APPROVED** [review]
🏆🏆🏆

File: Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift [hunk 7]
```diff
@@ -137,50 +137,56 @@ fileprivate extension TrialOrIntroPriceEligibilityChecker {
...
+            self.getIntroEligibilityFromBackend(with: receiptData,
+                                                productIdentifiers: productIdentifiers,
+                                                completion: completion)
         }
     }
 
 }
 
-private extension TrialOrIntroPriceEligibilityChecker {
+internal extension TrialOrIntroPriceEligibilityChecker {
```
[the remark this answers is not in the document]
Comment (joshdholtz) [reply Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift:164]: Ah okay! I can remove it 😊 I was playing with the options here to see what worked to let me test it 😛
